### PR TITLE
feat(directory): add studies associated to collections

### PR DESCRIPTION
**Pull request by svituz**
### What are the main changes you did:

I added, to the Biobank Directory module, the possibility of linking a clinical study to one or more collections. The clinical study is the context in which the collection of samples was generated. This PR is needed to link BBMRI ERIC collections with [ECRIN MDR](https://newmdr.ecrin.org/) clinical studies

The PR implements the following changes:
  * Model:
    * adds the `Study` entity with some attributes for the studies
    * adds to the `Collection` entity the attribute `study` that references a study
    * changes the attribute `national_node` of the `AlsoKnowIn` entity, to be optional. Indeed, in case the `also_known` entity is used to store information for a study, the national node is null, since the concept is not related to a study
  * UI
    * adds to the Collection page, in the info card on the right, a link to the details of the Study, if present
    * adds a Study page with the information of the Study. The page contains study attributes, the list of collections with the samples of the study, and an info card that just contains the link to the external catalog if present

### how to test:

- As the usual biobank_directory:
   * run molgenis-emx2
   * create a new database named Directory, using "Biobank_Directory" as a template setting "load data" to true
   * navigate to the details of the first collection of Biobank 1 (Collection 1 of biobank 1)
   * in the info card on the right there is a "Study" section with a link that brings to the Study details page

todo:
- [ ] updated docs in case of new feature
- [ ] added/updated tests
- [ ] added/updated testplan to include a test for this fix, including ref to bug using # notation

### CODE REVIEW

**Review by esthervanenckevort — CHANGES_REQUESTED** [review]
Please implement a feature flag to make it possible to enable / disable this feature

**Comment by esthervanenckevort** [comment]
The data model should also include a National Node column for studies. @dtroelofsprins can explain.

**Comment by dtroelofsprins** [comment]
Discussed this with Esther.
- In your comments you say that studies can contain collections from multiple national nodes and therefore no NN can be attached to it. We have similar cases with 'inter'national Networks, and these are included in a EU national node. We would like to propose to handle the studies in a similar way, so then the required=false in the AlsoKnownIN can be true again and a national_node column should be added to the Studies table.
- Second argument for this is, that it can be that national nodes want to include studies by themselves and that cannot be done on the proposed Studies table itself, but needs to go through a NN's own studies table which are copied into the ERIC studies tables overnight => and for this the NN column is relevant. 
- As a result of this I think it might also be good to broaden the description of the study ID in the molgenis.csv to a more general one as it's now only focussing on ECRIN. 
- And related to that: In your example you have id: bbmri_eric:ID:DE:112233, I would suggest to keep this in line with the IDs of the other tables, so something like: bbmri-eric:studyID:DE_112233

File: apps/directory/src/stores/studyStore.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      (column) => column.type === "range"
+    );
+
+    for (const property of rangeProperties) {
+      properties.push(property.min, property.max, property.unit_column);
+    }
+
+    return properties;
+  }
+
```
Comment (chinook25) [inline apps/directory/src/stores/studyStore.js:26]: Please split this out to a a util file and write a unittest for it.

File: apps/directory/src/functions/viewmodelMapper.js [hunk 2]
```diff
@@ -357,6 +375,14 @@ export const collectionReportInformation = (collection) => {
   return collectionReport;
 };
 
+export const studyReportInformation = (study) => {
```
Comment (chinook25) [inline apps/directory/src/functions/viewmodelMapper.js:378]: Rename to getStudyReportInformation (or something like it), currently it read like a variable name.

File: apps/directory/src/views/StudyReport.vue [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+});
+
+const uiText = computed(() => settingsStore.uiText);
+
+const studyDataAvailable = computed(() => {
+  return Object.keys(study).length;
+});
+
+const info = computed(() => {
+  return studyDataAvailable.value ? studyReportInformation(study.value) : {};
```
Comment (chinook25) [inline apps/directory/src/views/StudyReport.vue:75]: studyReportInformation always returns, so ternary is not needed.

File: apps/directory/src/views/StudyReport.vue [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+});
+
+function loadStudyReport(id) {
+  loaded.value = false;
+  const studyPromise = studyStore.getStudyReport(id).then((result) => {
+    study.value = result.Studies.length ? result.Studies[0] : {};
+  });
+  Promise.all([studyPromise]).then(() => {
```
Comment (chinook25) [inline apps/directory/src/views/StudyReport.vue:83]: `promise.all` is not need, can just do `studyPromise.then...`

Comment (chinook25) [reply apps/directory/src/views/StudyReport.vue:83]: Or even remove it and combine it with the body of the .then above it.

File: apps/directory/src/views/StudyReport.vue [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+const studyStore = useStudyStore();
+
+const route = useRoute();
+const study = ref({});
+
+let loaded = ref(false);
+
+loadStudyReport(route.params.id);
+
+watch(route, async (route) => {
```
Comment (chinook25) [inline apps/directory/src/views/StudyReport.vue:64]: Does this need to be a deep watch?

File: apps/directory/src/components/report-components/StudyReportInfoCard.vue [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    </div>
+  </div>
+</template>
+
+<script setup>
+import { toRefs } from "vue";
+import ReportDetailsList from "../../components/report-components/ReportDetailsList.vue";
+
+const props = defineProps(["info"]);
+let { info } = toRefs(props);
```
Comment (chinook25) [inline apps/directory/src/components/report-components/StudyReportInfoCard.vue:21]: const

File: apps/directory/src/views/StudyReport.vue [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  loadStudyReport(route.params.id);
+});
+
+const uiText = computed(() => settingsStore.uiText);
+
+const studyDataAvailable = computed(() => {
```
Comment (chinook25) [inline apps/directory/src/views/StudyReport.vue:70]: variable seems unused?

**Review by chinook25 — CHANGES_REQUESTED** [review]
Please make sure new files are written in typescript and for .vue files use the setup tag.

File: apps/directory/src/views/StudyReport.vue (hunk 4 above)
Comment (svituz) [reply apps/directory/src/views/StudyReport.vue:83]: added to the first .then

File: apps/directory/src/views/StudyReport.vue (hunk 3 above)
Comment (svituz) [reply apps/directory/src/views/StudyReport.vue:75]: removed ternary operator

File: apps/directory/src/views/StudyReport.vue (hunk 7 above)
Comment (svituz) [reply apps/directory/src/views/StudyReport.vue:70]: removed the variable

File: apps/directory/src/functions/viewmodelMapper.js (hunk 2 above)
Comment (svituz) [reply apps/directory/src/functions/viewmodelMapper.js:378]: done

File: apps/directory/src/property-config/initialCollectionColumns.js [hunk 8]
```diff
@@ -78,6 +78,11 @@ const initialCollectionColumns = [
     column: { data_use: ["label", "ontologyTermURI"] },
     type: "array",
   },
+  {
+    label: "Study:",
+    column: { study: ["id", "title"] },
+    type: "xref",
```
[the remark this answers is not in the document]
Comment (svituz) [reply apps/directory/src/property-config/initialCollectionColumns.js:84]: replaced with object

File: apps/directory/src/components/report-components/StudyReportInfoCard.vue (hunk 6 above)
Comment (svituz) [reply apps/directory/src/components/report-components/StudyReportInfoCard.vue:21]: done

File: apps/directory/src/views/StudyReport.vue (hunk 5 above)
Comment (svituz) [reply apps/directory/src/views/StudyReport.vue]: Not indeed, It needs just to be loaded once. I removed it

File: apps/directory/src/stores/studyStore.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+  async function getStudyReport(id: string) {
+    console.log(getStudyColumns());
+    const studyReportQuery = new QueryEMX2(graphqlEndpoint)
+      .table("Studies")
+      .select(getStudyColumns())
+      .orderBy("Studies", "id", "asc")
+      .where("id")
+      .like(id);
+    console.log(studyReportQuery);
```
Comment (chinook25) [inline apps/directory/src/stores/studyStore.ts:34]: also remove

**Review by chinook25 — APPROVED** [review]
Works for me! Just tidy up the code a bit (see comments) and I think it's done.

**Review by dtroelofsprins — CHANGES_REQUESTED** [review]
Thanks for making all the adjustment after our first review, in general data model etc looks fine now, except from some small remarks => see comments in the files.

Also the app looks fine:
- Link to study card in collection detail view works OK.
- Separate study card looks fine => All information that can be entered in the studies table is displayed if available and if not then it's also not displayed.

One small remark: why is study id displayed in the details on the collection card itself?
image. Because the information is also available via the right pane? Like f.e. biobanks and for networks these IDs are also not in the collection details.

Second thing is more a feature request and I have no idea if it's difficult to implement, but I think it might be nice if you could navigate from the study card back to the collection card. So that the breadcrumb gets similar to the one of the collection card:
image
But then "Back to the catalogue"/"Collection1 of biobank1"/Dummy study? Instead of "Back to the catalogue"/Dummy study?

File: apps/directory/vite.config.js [hunk 10]
```diff
@@ -5,7 +5,7 @@ import { createHtmlPlugin } from "vite-plugin-html";
 import monacoEditorPlugin from "vite-plugin-monaco-editor";
 
 const HOST =
-  process.env.MOLGENIS_APPS_HOST || "https://bbmri-emx2-test.molgeniscloud.org";
+  process.env.MOLGENIS_APPS_HOST || "http://localhost:8080/" // "https://bbmri-emx2-test.molgeniscloud.org";
```
[the remark this answers is not in the document]
Comment (svituz) [reply apps/directory/vite.config.js]: done

File: apps/directory/src/stores/studyStore.ts [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+      (column) => column.type === "range"
+    );
+    for (const property of rangeProperties) {
+      properties.push(property.min, property.max, property.unit_column);
+    }
+    return properties;
+  }
+
+  async function getStudyReport(id: string) {
+    console.log(getStudyColumns());
```
[the remark this answers is not in the document]
Comment (svituz) [reply apps/directory/src/stores/studyStore.ts]: removed

File: apps/directory/src/stores/studyStore.ts (hunk 9 above)
Comment (svituz) [reply apps/directory/src/stores/studyStore.ts]: removed

File: data/biobank-directory/molgenis.csv [hunk 12]
```diff
@@ -161,4 +162,17 @@ Persons,,role,Role of the Person,text,,false,,Official role of the person.,,,,,,
 Persons,,biobanks,Biobanks,refback,,false,true,Biobanks the person is connected to.,,Biobanks,contact,,,,,,
 Persons,,collections,Collections,refback,,false,true,Collections the person is connected to.,,Collections,contact,,,,,,
 Persons,,networks,Networks,refback,,false,true,Networks the person is connected to.,,Networks,contact,,,,,,
-Persons,,national_node,National Node,ref,,true,,The person originates from this national node.,,NationalNodes,,${description},,,,,
\ No newline at end of file
+Persons,,national_node,National Node,ref,,true,,The person originates from this national node.,,NationalNodes,,${description},,,,,
+Studies,,,,,,,,Table with information about the studies in which context the collection was generated,,,,,,,,,
+Studies,,id,ID,string,1,true,,"Unique ID of the study, prefixed with bbmri-eric:ID:<ecrin_mdr_id>",,,,,,,,,
```
[the remark this answers is not in the document]
Comment (svituz) [reply data/biobank-directory/molgenis.csv]: removed all ECRIN reference in the model

File: data/biobank-directory/molgenis.csv [hunk 13]
```diff
@@ -115,6 +115,7 @@ Collections,,access_joint_project,Access via Join Projects to,ontology_array,,fa
 Collections,,access_description,Access Description,text,,false,,Short description of the access rules.,,,,,,,,,
 Collections,,access_uri,Access URI,hyperlink,,false,,URI describing the access policy.,,,,,,,,,
 Collections,,sop,PD/SOPs,ontology_array,,false,,Availability of Process Descriptions (PDs) and/or Standard Operating Procedures (SOPs).,DirectoryOntologies,SOPs,,${label},,,,,
+Collections,,study,Study,ref,,false,,A link to a Study during which the collection was generated,,Studies,,,,,,,
```
[the remark this answers is not in the document]
Comment (svituz) [reply data/biobank-directory/molgenis.csv]: added

**Comment by svituz** [comment]
@dtroelofsprins I removed the study from the card details. You were right it was unnecessary. 
Regarding the breadcrumbs, I thought about adding the collection to the path. The problem, though, is that the relation Study-Collections is 1-N, so a study can be associated with multiple collections. When navigating from the collection page it is possible to know the collection but this is not possible when accessing the page via direct URL. So to keep the same behavior in the two cases I preferred not to include the breadcrumb in all situations.
The same thing happens for the Networks.

**Comment by dtroelofsprins** [comment]
> @dtroelofsprins I removed the study from the card details. You were right it was unnecessary. Regarding the breadcrumbs, I thought about adding the collection to the path. The problem, though, is that the relation Study-Collections is 1-N, so a study can be associated with multiple collections. When navigating from the collection page it is possible to know the collection but this is not possible when accessing the page via direct URL. So to keep the same behavior in the two cases I preferred not to include the breadcrumb in all situations. The same thing happens for the Networks.

Ah of course you're right about that, studies can have more collections, which make me wonder: currently a collection can only be in one study, is this indeed true? Or could it potentially be linked to multiple studies?

File: data/biobank-directory/molgenis.csv [hunk 14]
```diff
@@ -115,6 +115,7 @@ Collections,,access_joint_project,Access via Join Projects to,ontology_array,,fa
 Collections,,access_description,Access Description,text,,false,,Short description of the access rules.,,,,,,,,,
 Collections,,access_uri,Access URI,hyperlink,,false,,URI describing the access policy.,,,,,,,,,
 Collections,,sop,PD/SOPs,ontology_array,,false,,Availability of Process Descriptions (PDs) and/or Standard Operating Procedures (SOPs).,DirectoryOntologies,SOPs,,${label},,,,,
+Collections,,study,Study,ref,,false,,A link to a Study during which the collection was generated,,Studies,,${title},,,,,
```
Comment (dtroelofsprins) [inline data/biobank-directory/molgenis.csv:118]: Sorry, forgot to mention it earlier, but now the study is shown under the Policies header. I think it's better to move this to the "Belongs to' one. So after also_known.

**Comment by esthervanenckevort** [comment]
> Ah of course you're right about that, studies can have more collections, which make me wonder: currently a collection can only be in one study, is this indeed true? Or could it potentially be linked to multiple studies?

A collection can be linked to multiple studies

**Comment by esthervanenckevort** [comment]
Can you have a look at how you can harmonise the data model with MIABIS v3 Study (researchresource).

https://docs.google.com/spreadsheets/d/1RrNHSNWotFl0F4P3u2cb-o_z-20bJGUn/edit?gid=1358611155#gid=1358611155

File: data/biobank-directory/molgenis.csv (hunk 14 above)
Comment (svituz) [reply data/biobank-directory/molgenis.csv]: moved

**Comment by svituz** [comment]
@esthervanenckevort The harmonization with MIABIS 3 needs further discussions with the MIABIS team, so it was decided to use the current model to start linking the MDR with the Directory and leave the harmonization for future work.

**Comment by dtroelofsprins** [comment]
> > Ah of course you're right about that, studies can have more collections, which make me wonder: currently a collection can only be in one study, is this indeed true? Or could it potentially be linked to multiple studies?
> 
> A collection can be linked to multiple studies

If so, then the datatype should be ref_array instead of ref

**Comment by svituz** [comment]
I changed the datatype of studies to ref_array as suggested and changed the frontend to support multiple studies

**Comment by dtroelofsprins** [comment]
> I changed the datatype of studies to ref_array as suggested and changed the frontend to support multiple studies

Thanks! 
One (last) small remark, as multiple studies now can be linked to the collections it would perhaps be better to name the column studies instead of study in the collections table?

**Comment by svituz** [comment]
@dtroelofsprins can you check this again?

**Comment by dtroelofsprins** [comment]
> @dtroelofsprins can you check this again?

I'm a bit confused now, not sure whether you've seen the comment I made last week?
> Thanks!
> One (last) small remark, as multiple studies now can be linked to the collections it would perhaps be better to name the > > column studies instead of study in the collections table?

Looking at the molgenis.csv the column in collections.csv is still named Study, instead of Studies, but I do see that new commits were pushed recently.

**Comment by svituz** [comment]
> > @dtroelofsprins can you check this again?
> 
> I'm a bit confused now, not sure whether you've seen the comment I made last week?
> 
> > Thanks!
> > One (last) small remark, as multiple studies now can be linked to the collections it would perhaps be better to name the > > column studies instead of study in the collections table?
> 
> Looking at the molgenis.csv the column in collections.csv is still named Study, instead of Studies, but I do see that new commits were pushed recently.

I must have missed the comment. I changed the name of the column to studies now

**Review by dtroelofsprins — APPROVED** [review]
Thanks for the latest change. 
As far as I can see all things look and work fine now :slightly_smiling_face:

**Pull request merged**